frame/staking/src/offchain_election.rs

```diff
@@ -155,19 +163,163 @@ pub fn get_balancing_iters<T: Trait>() -> usize {
 	}
 }
 
+/// Find the maximum `len` that a compact can have in order to fit into the block weight.
+///
+/// We start by assuming `voters == size.nominators`, i.e. all nominators are active in the
+/// solution. Then we do a binary search until we find the ideal value.
+///
+/// Thus, this only returns a value between zero and `size.nominators`.
+pub fn maximum_compact_len<W: crate::WeightInfo>(
+	winners_len: u32,
+	size: ElectionSize,
+	max_weight: Weight,
+) -> u32 {
+	use sp_std::cmp::Ordering;
+	if size.nominators < 1 {
+		return size.nominators;
+	}
+
+	let weight_with = |voters: u32| -> Weight {
+		W::submit_solution_better(
+			size.validators.into(),
+			size.nominators.into(),
+			voters,
+			winners_len,
+		)
+	};
+
+	let mut voters = size.nominators.max(2);
+	let mut step = voters;
+	loop {
+		let current = weight_with(voters);
+		step = step / 2;
+		if step == 0 {
+			// We might have reduced less than expected due to rounding error. Reduce linearly one
+			// last time.
+			while weight_with(voters) > max_weight {
+				if let Some(next) = voters.checked_sub(1) {
+					voters = next
+				} else {
+					break;
+				}
+			}
+			// For the opposite case.
+			while weight_with(voters + 1) < max_weight {
+				if voters + 1 >= size.nominators {
+					break
+				}
```

Comment:
```suggestion
			while votes <= size.nominators && weight_with(voters + 1) < max_weight {
```

I believe `voters == size.nominators` is still fine.